tests/framework/ab_test.py

```diff
@@ -0,0 +1,130 @@
+    to `cargo_build.get_binary`).
+
+    Note that there are no guarantees on the order in which the two tests are run.
+
+    :param test_runner: A callable which when executed runs the test in the context of the current working directory. Its
+                        parameter is `true` if and only if it is currently running the "A" test.
+    :param comparator: A callable taking two outputs from `test_runner` and comparing them. Should return some value
+                       indicating whether the test should pass or no, which will be returned by the `ab_test` functions,
+                       and on which the caller can then do an assertion.
+    :param a_revision: The revision to checkout for the "A" part of the test. Defaults to the pull request target branch
+                       if run in CI, and "main" otherwise.
+    :param b_revision: The git revision to check out for "B" part of the test. Defaults to whatever is currently checked
+                       out (in which case no temporary directory will be created).
+    :return: The output of both "A" test, the "B" test and the comparator, which can then be used for assertions
+             (alternatively, your comparator can perform any required assertions and not return anything).
+    """
+
+    # We can't just checkout random branches in the current working directory. Locally, this might not work because of
+    # uncommitted changes. In the CI this will not work because multiple tests will run in parallel, and thus switching
+    # branches will cause random failures in other tests.
+    with temporary_checkout(a_revision) as a_tmp:
+        with chdir(a_tmp):
+            result_a = test_runner(True)
+
+        if b_revision:
+            with temporary_checkout(b_revision) as b_tmp:
+                with chdir(b_tmp):
+                    result_b = test_runner(False)
+                # Have to call comparator here to make sure both temporary directories exist (as the comparator
+                # might rely on some files that were created during test running, see the benchmark test)
+                comparison = comparator(result_a, result_b)
+        else:
+            # By default, pytest execution happens inside the `tests` subdirectory. Change to the repository root, as
+            # documented.
+            with chdir(".."):
+                result_b = test_runner(False)
+            comparison = comparator(result_a, result_b)
+
+        return result_a, result_b, comparison
+
+
+@contextlib.contextmanager
+def temporary_checkout(revision: str):
+    """
+    Context manager that checks out firecracker in a temporary directory and `chdir`s into it
+
+    Will change back to whatever was the current directory when the context manager was entered, even if exceptions
+    happen along the way.
+    """
+    with TemporaryDirectory() as tmp_dir:
+        utils.run_cmd(
+            f"git clone https://github.com/firecracker-microvm/firecracker {tmp_dir}"
```

Comment:
Are there any situations where we wouldn't want this hardcoded?

Reply:
Yeah, for example if someone wants to run an A/B-test against a commit that's not in the firecracker repository (say, because its in a fork). Since right now this is only used for testing in PRs (where A = head of PR target inside firecracker repo and B = whatever is currently checked out), this is not a limitation we need to address just now though